### PR TITLE
Clean up the broken Class and Function links in docstrings

### PR DESCRIPTION
**Pull request by yck011522**
### What type of change is this?

- [x] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [x] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas_fab.robots.CollisionMesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas_fab/robots/planning_scene.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def __data__(self):
 
 
 class AttachedCollisionMesh(Data):
-    """Represents a collision mesh that is attached to a :class:`Robot`'s :class:`~compas_robots.model.Link`.
+    """Represents a collision mesh that is attached to a :class:`Robot`'s :class:`compas_robots.model.Link`.
```
Comment (gonzalocasas) [inline src/compas_fab/robots/planning_scene.py:122]: I actually like this with the `~` better, otherwise it is too verbose, don't you think?

File: src/compas_fab/robots/robot.py [hunk 2]
```diff
@@ -31,21 +31,24 @@ class Robot(Data):
     upon the model described in the class :class:`compas_robots.RobotModel` of
     the **COMPAS** framework.
 
-    Attributes
+    Parameters
     ----------
-    model : :class:`RobotModel`
+    model : :class:`compas_robots.RobotModel`
```
Comment (gonzalocasas) [inline src/compas_fab/robots/robot.py:36]: You don't need to specify the package if it's already imported. Unless you're doing this intentionally to highlight that the model is coming from another library?

File: src/compas_fab/robots/robot.py [hunk 3]
```diff
@@ -31,21 +31,24 @@ class Robot(Data):
...
+    semantics : :class:`~compas_fab.robots.RobotSemantics`
         The semantic model of the robot. Defaults to ``None``.
-    client : :class:`compas_fab.backends.interfaces.ClientInterface`
+    client : :class:`~compas_fab.backends.interfaces.ClientInterface`
         The backend client to use for communication,
-        e.g. :class:`compas_fab.backends.RosClient`
+        e.g. :class:`~compas_fab.backends.RosClient`
+
+    Attributes
+    ----------
```
Comment (gonzalocasas) [inline src/compas_fab/robots/robot.py:47]: I agree with documenting `Parameters`, but now this part of `Attribuets` is incomplete now, because it only lists attributes not present in the constructor's part.

File: src/compas_fab/robots/semantics.py [hunk 4]
```diff
@@ -15,6 +15,49 @@ class RobotSemantics(Data):
...
+        The order is not important, i.e. the pair `('link1', 'link2')` is the same as `('link2', 'link1')`.
+        Only one pair is needed.
+    group_states : dict, optional
+        A nested dictionary defining named states for a particular group, in terms of joint values.
+        This is useful to define states such as "home" or "folded arms" for the robot.
+        The dictionary structure is as follows:
+
+        - Level 1 keys are planning group names : :obj:`str`.
+        - Level 2 keys are group state names : :obj:`str`.
+        - Level 3 keys are joint names and values are joint values : :obj:`str`.
```
Comment (gonzalocasas) [inline src/compas_fab/robots/semantics.py:53]: Maybe add an example instead of the bullet-point description? Or both?

File: src/compas_fab/robots/semantics.py [hunk 5]
```diff
@@ -122,23 +184,78 @@ def from_xml(cls, xml, robot_model):
         )
 
     def get_end_effector_link_name(self, group=None):
+        """Get the :obj:`str` name of the last link (end effector link) in a planning group.
```
Comment (gonzalocasas) [inline src/compas_fab/robots/semantics.py:187]:
```suggestion
        """Get the name of the last link (end effector link) in a planning group.
```
doesn't it read weird to see ```Get the `str` name of the ...```? A name is a string anyway

File: src/compas_fab/robots/semantics.py [hunk 6]
```diff
@@ -122,23 +184,78 @@ def from_xml(cls, xml, robot_model):
...
+        """
         if not group:
             group = self.main_group_name
         return self.groups[group]["links"][0]
 
     def get_all_configurable_joints(self):
+        """Get all configurable :class:`compas_robots.model.Joint` of the robot.
+
+        Configurable joints are joints that can be controlled,
+        aka, not ``Joint.Fixed``, not mimicking another joint and not a passive joint.
```
Comment (gonzalocasas) [inline src/compas_fab/robots/semantics.py:225]:
```suggestion
        aka, not ``Joint.FIXED``, not mimicking another joint and not a passive joint.
```

File: src/compas_fab/robots/planning_scene.py (hunk 1 above)
Comment (yck011522) [reply src/compas_fab/robots/planning_scene.py:122]: Yea. But is it clear to people that compas_robots is a different package? At the moment if you pretend you are a user following that clickable link, you will reach a page that simply list the class Link. Somehow misleading that the Link class belongs to compas_fab.

I think if the API documentation page theme is slightly different between compas_fab and compas_robots, (maybe a different colour) then we can go with ~ for all the compas_robot links.

File: src/compas_fab/robots/robot.py (hunk 2 above)
Comment (yck011522) [reply src/compas_fab/robots/robot.py:36]: Yea
 That was the intention, see earlier reply

File: src/compas_fab/robots/robot.py (hunk 3 above)
Comment (yck011522) [reply src/compas_fab/robots/robot.py:47]: So I guess some of the attributes will be a repeat of the parameters. Yea?

File: src/compas_fab/robots/semantics.py (hunk 5 above)
Comment (yck011522) [reply src/compas_fab/robots/semantics.py:187]: I guess a name is a str... Okay.

File: src/compas_fab/robots/tool.py [hunk 7]
```diff
@@ -15,19 +15,32 @@
...
     """Represents a tool to be attached to the robot's flange.
 
-    Attributes
+    Properties
     ----------
     visual : :class:`compas.datastructures.Mesh`
         The visual mesh of the tool.
     frame_in_tool0_frame : :class:`compas.geometry.Frame`
-        The frame of the tool in tool0 frame.
+        The tool coordinate frame (TCF) of the tool relative to the planner coordinate frame (PCF).
```
Comment (gonzalocasas) [inline src/compas_fab/robots/tool.py:23]: Please don't introduce new frame names in this PR, this is about a much simpler issue.

File: src/compas_fab/robots/time_.py [hunk 8]
```diff
@@ -30,6 +52,11 @@ def __init__(self, secs, nsecs):
         self.secs = int(quotient)
         self.nsecs = int(remainder * sec_to_nano_factor) + int(nsecs)
 
+        # If nsecs is greater than 1 second, add the remainder back to secs
+        if self.nsecs >= sec_to_nano_factor:
+            self.secs += 1
+            self.nsecs -= int(sec_to_nano_factor)
+
```
Comment (gonzalocasas) [inline src/compas_fab/robots/time_.py:59]: Mmmm.. ok, but please move to a separate PR

File: src/compas_fab/robots/tool.py (hunk 7 above)
Comment (yck011522) [reply src/compas_fab/robots/tool.py:23]: Okay, changed back to T0CF

File: src/compas_fab/robots/robot.py (hunk 3 above)
Comment (yck011522) [reply src/compas_fab/robots/robot.py:47]: Fixed.

File: src/compas_fab/robots/semantics.py (hunk 4 above)
Comment (yck011522) [reply src/compas_fab/robots/semantics.py:53]: It is very involved in explaining this properly. I think these bullet points is more of a note now.
Maybe we can do the example in a separate rst doc in tutorials.

**Comment by yck011522** [comment]
@gonzalocasas Fixed all your comments, we need to discuss only one thing. Whether to use the ~ for external links that points to compas_robots

File: src/compas_fab/robots/planning_scene.py (hunk 1 above)
Comment (gonzalocasas) [reply src/compas_fab/robots/planning_scene.py:122]: You're right, it's useful to keep the full link in these cases, to highlight it's an external module. Altering the colors of fab would be a bit against the idea of having homogenous docs for all compas, so, let's stick with full link

File: src/compas_fab/robots/robot.py [hunk 9]
```diff
@@ -1983,7 +2024,10 @@ def scale(self, factor):
 
     @property
     def scale_factor(self):
-        """:obj:`float`: Robot's scale factor."""
+        """A scale factor affecting planning targets, results and visualizaion; Typically, robot models are defined in meters,
```
Comment (gonzalocasas) [inline src/compas_fab/robots/robot.py:2027]:
```suggestion
        """A scale factor affecting planning targets, results and visualization. Typically, robot models are defined in meters,
```

File: src/compas_fab/robots/semantics.py [hunk 10]
```diff
@@ -122,23 +180,78 @@ def from_xml(cls, xml, robot_model):
...
+        """
         if not group:
             group = self.main_group_name
         return self.groups[group]["links"][0]
 
     def get_all_configurable_joints(self):
+        """Get all configurable :class:`compas_robots.model.Joint` of the robot.
+
+        Configurable joints are joints that can be controlled,
+        aka, not ``Joint.FIXED``, not mimicking another joint and not a passive joint.
```
Comment (gonzalocasas) [inline src/compas_fab/robots/semantics.py:221]:
```suggestion
        i.e., not ``Joint.FIXED``, not mimicking another joint and not a passive joint.
```

File: src/compas_fab/robots/semantics.py [hunk 11]
```diff
@@ -122,23 +180,78 @@ def from_xml(cls, xml, robot_model):
...
         for joint in self.robot_model.get_configurable_joints():
             if joint.name not in self.passive_joints:
                 joints.append(joint)
         return joints
 
     def get_configurable_joints(self, group=None):
+        """Get all configurable :class:`compas_robots.model.Joint` of a planning group.
+
+        Configurable joints are joints that can be controlled,
+        aka, not ``Joint.FIXED``, not mimicking another joint and not a passive joint.
```
Comment (gonzalocasas) [inline src/compas_fab/robots/semantics.py:241]:
```suggestion
        i.e., not ``Joint.FIXED``, not mimicking another joint and not a passive joint.
```

**Review by gonzalocasas — APPROVED** [review]
Looking good! I just added a couple of suggestions for typos, otherwise, ready to go!
